src/library_webgl.js

```diff
@@ -4021,8 +4021,18 @@ var LibraryGL = {
   glMapBufferRange__sig: 'iiiii',
   glMapBufferRange__deps: ['$emscriptenWebGLGetBufferBinding', '$emscriptenWebGLValidateMapBufferTarget'],
   glMapBufferRange: function(target, offset, length, access) {
-    if (access != 0x1A && access != 0xA) {
-      err("glMapBufferRange is only supported when access is MAP_WRITE|INVALIDATE_BUFFER");
+    if ((access & (0x1 | 0x20)) != 0) {
+      err("glMapBufferRange access does not support MAP_READ or MAP_UNSYNCHRONIZED");
+      return 0;
+    }
+
+    if ((access & 0x2) == 0) {
+      err("glMapBufferRange access must include MAP_WRITE");
+      return 0;
+    }
+
+    if ((access & (0x4 | 0x8)) == 0) {
+      err("glMapBufferRange access must include INVALIDATE_BUFFER or INVALIDATE_RANGE");
```

Comment:
Something I find here already wrong with the old code is that these would probably be better to generate GL errors, but instead they just log an error and return. Adding a call to `GL.recordError(0x501 /* GL_INVALID_VALUE */);` in all of these paths would be ok, e.g.

```
   if ((access & (0x1 | 0x20)) != 0) {
      err("glMapBufferRange access does not support MAP_READ or MAP_UNSYNCHRONIZED");
      GL.recordError(0x501 /* GL_INVALID_VALUE */);
      return 0;
   }
```

Also, it might be nice to inline a comment for the raw bit values, something like
```
 if ((access & (0x1/*GL_MAP_READ_BIT*/ | 0x20/*GL_MAP_UNSYNCHRONIZED_BIT*/)) != 0) {
```

Neither of these are particularly critical though.

Reply:
I wanted to generate GL errors as well.  I started to add the code to do that, but remembered a typical OpenGL problem--there is no "GL_UNSUPPORTED_OPERATION" or any kind of error code to indicate that the implementation is limited beyond the scope of the spec.  I've seen cases where GL drivers just raise GL_INVALID_OPERATION when they don't support something, but then when I go to look at the OpenGL docs to try to debug it, I'm confused and lost because the error "shouldn't be happening" according to the spec.  Only some lucky googling or guessing saves me.  What about throwing some kind of javascript exception so the user hopefully knows to go look in the debug console?  The console warnings about needing INVALIDATE_BUFFER were what saved me in the first place.

Good call on the constant comment, I'll add that.

Reply:
Comments done.  Let me know if you'd like GL errors or what.  I do wish there were some sort of GL_INTERNAL_ERROR or GL_UNIMPLEMENTED or what-have-you...  The super nice thing would be to implement glDebugMessageCallback() which is in KHR_debug and became core in ES 3.2.

Reply:
Hello, friendly ping.

Reply:
Yeah, agree that GL error management is not particularly capable. Looks good.